File: analysis/05_soccer/soccer_loop.R
# The soccer analysis loop: does Elo-vs-DraftKings 3-way earn anything, and are the
# soccer constants in soccer_edge.py the right priors?
#
#   A. Paper ROI by pick x strength, flat $1, bootstrap 95% CI.
#   B. Slices a rule can act on: edge band, price band, pick (Wilson CI on hit, flat ROI).
#   C. Elo calibration on settled snapshots + 3-way log-loss of the model vs the closer.
#   D. Refit on the results table: replay Elo for a grid of ELO_HFA, score every DRAW_BASE
#      by 3-way log-likelihood on the held-out second half.
#
# Output: analysis/_out/soccer_*.csv. Mirrors soccer_loop.py — keep in lockstep.

suppressPackageStartupMessages({
  library(dplyr)
  library(boot)
})

.this_dir <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) "analysis/05_soccer")
source(file.path(.this_dir, "..", "_shared", "load_data.R"))     # wilson()
source(file.path(.this_dir, "..", "_shared", "load_soccer.R"))

OUT_DIR   <- normalizePath(file.path(.this_dir, "..", "_out"), mustWork = FALSE)
MIN_BETS  <- 10
MIN_N     <- 5
BOOT_REPS <- 5000
SEED      <- 20260920

ELO_START <- 1500; ELO_K <- 20; ELO_K_FRIENDLY <- 10
ELO_HFA_LIVE <- 60; DRAW_BASE_LIVE <- 0.26
HFA_GRID  <- c(0, 30, 60, 90, 120)
DRAW_GRID <- c(0.20, 0.22, 0.24, 0.26, 0.28, 0.30, 0.32)
HOLDOUT_FRAC <- 0.5

EDGE_BINS  <- c(0, 8, 15, 20, 30, 50, 100000)
EDGE_LABELS <- c("0-8", "8-15", "15-20", "20-30", "30-50", "50+")
PRICE_BINS <- c(-100000, -200, -110, 99, 150, 200, 250, 100000)
PRICE_LABELS <- c("<-200", "-200..-110", "-110..+100", "+100..150", "+150..200", "+200..250", ">+250")
PROB_BINS  <- c(0, .2, .3, .4, .5, .6, .7, .8, 1.01)
PROB_LABELS <- c("<.2", ".2-.3", ".3-.4", ".4-.5", ".5-.6", ".6-.7", ".7-.8", ".8+")

dir.create(OUT_DIR, showWarnings = FALSE, recursive = TRUE)
set.seed(SEED)

verdict <- function(lo, hi) if (lo > 0) "PROFITABLE (95% CI > 0)" else if (hi < 0) "losing (95% CI < 0)" else "inconclusive"
boot_ci <- function(x) {
  b <- boot(x, function(d, i) mean(d[i]), R = BOOT_REPS)
  ci <- boot.ci(b, type = "perc")$percent
  c(lo = ci[4], hi = ci[5])
}

# ---------------------------------------------------------------- A
section_a <- function(bets) {
  cat(sprintf("A. Paper ROI — %s settled soccer paper bets, backfilled %s, flat ROI %+.1f%%\n\n",
              format(nrow(bets), big.mark = ","), format(sum(bets$backfill), big.mark = ","), 100 * mean(bets$pnl_flat)))
  groups <- list(list(pick = "ALL", strength = "all", g = bets))
  for (key in split(bets, list(bets$pick, bets$strength), drop = TRUE))
    groups[[length(groups) + 1]] <- list(pick = key$pick[1], strength = as.character(key$strength[1]), g = key)
  for (key in split(bets, bets$pick))
    groups[[length(groups) + 1]] <- list(pick = key$pick[1], strength = "any", g = key)
  for (key in split(bets, bets$strength))
    groups[[length(groups) + 1]] <- list(pick = "ALL", strength = as.character(key$strength[1]), g = key)
  rows <- lapply(groups, function(x) {
    g <- x$g
    if (nrow(g) < MIN_BETS) return(NULL)
    ci <- boot_ci(g$pnl_flat)
    data.frame(pick = x$pick, strength = x$strength, bets = nrow(g), wins = sum(g$result == "W"),
               hit_rate = mean(g$result == "W"), roi_mean = mean(g$pnl_flat),
               roi_ci_lo = ci[["lo"]], roi_ci_hi = ci[["hi"]], verdict = verdict(ci[["lo"]], ci[["hi"]]),
               stringsAsFactors = FALSE)
  })
  out <- bind_rows(rows) %>% arrange(desc(roi_ci_lo))
  cat(sprintf("  %-7s%4s%6s%6s%7s%8s%8s%8s  verdict\n", "pick", "str", "bets", "wins", "hit%", "ROI", "CI lo", "CI hi"))
  for (i in seq_len(nrow(out))) {
    r <- out[i, ]
    cat(sprintf("  %-7s%4s%6d%6d%6.1f%%%+7.1f%%%+7.1f%%%+7.1f%%  %s\n", r$pick, r$strength, r$bets, r$wins,
                100 * r$hit_rate, 100 * r$roi_mean, 100 * r$roi_ci_lo, 100 * r$roi_ci_hi, r$verdict))
  }
  out
}

# ---------------------------------------------------------------- B
slice_table <- function(df, col, order, title) {
  cat(sprintf("\n%s\n", title))
  cat(sprintf("  %-12s%5s%5s%8s%8s   95%% CI on hit\n", "bucket", "n", "W", "hit%", "ROI"))
  rows <- list()
  for (key in order[order %in% unique(df[[col]])]) {
    g <- df[df[[col]] == key, ]
    if (nrow(g) < MIN_N) next
    w <- sum(g$result == "W"); wi <- wilson(w, nrow(g))
    rows[[length(rows) + 1]] <- data.frame(dimension = col, bucket = key, n = nrow(g), wins = w, hit_rate = wi[1],
                                           hit_lo = wi[2], hit_hi = wi[3], roi_flat = mean(g$pnl_flat),
                                           stringsAsFactors = FALSE)
    cat(sprintf("  %-12s%5d%5d%7.1f%%%+7.1f%%   [%.0f, %.0f]\n", key, nrow(g), w, 100 * wi[1],
                100 * mean(g$pnl_flat), 100 * wi[2], 100 * wi[3]))
  }
  rows
}

section_b <- function(bets) {
  b <- bets[bets$strength >= 1, ]
  b$edge_band  <- as.character(cut(b$edge, EDGE_BINS, labels = EDGE_LABELS, right = FALSE))
  b$price_band <- as.character(cut(b$price, PRICE_BINS, labels = PRICE_LABELS, right = TRUE))
  rows <- slice_table(b, "edge_band", EDGE_LABELS, "B1. by edge band (does a bigger Elo-vs-DK gap win more?)")
  rows <- c(rows, slice_table(b, "price_band", PRICE_LABELS, "B2. by price band"))
  rows <- c(rows, slice_table(b, "pick", c("home", "draw", "away"), "B3. by pick"))
  bind_rows(rows)
}

# ---------------------------------------------------------------- C
section_c <- function(m) {
  cat(sprintf("\nC. Elo calibration on %s settled matches with a closer and a rating on both sides\n",
              format(nrow(m), big.mark = ",")))
  rows <- list()
  for (spec in list(c("home_p", "P(home win)", "home"), c("draw_p", "P(draw)", "draw"), c("away_p", "P(away win)", "away"))) {
    col <- spec[1]; hit <- spec[3]
    cat(sprintf("  %s: model bin -> observed rate (Wilson 95%%)\n", spec[2]))
    band <- as.character(cut(m[[col]], PROB_BINS, labels = PROB_LABELS, right = FALSE))
    for (key in PROB_LABELS[PROB_LABELS %in% unique(band)]) {
      g <- m[band == key, ]
      if (nrow(g) < MIN_N) next
      k <- sum(g$outcome == hit); wi <- wilson(k, nrow(g)); pred <- mean(g[[col]])
      rows[[length(rows) + 1]] <- data.frame(outcome = hit, bin = key, n = nrow(g), pred = pred, obs = wi[1],
                                             obs_lo = wi[2], obs_hi = wi[3], stringsAsFactors = FALSE)
      cat(sprintf("    %-7s n=%5d  pred %5.1f%%  obs %5.1f%% [%5.1f,%5.1f]  Δ%+5.1fpp\n", key, nrow(g),
                  100 * pred, 100 * wi[1], 100 * wi[2], 100 * wi[3], 100 * (wi[1] - pred)))
    }
  }
  eps <- 1e-9
  mp <- ifelse(m$outcome == "home", m$home_p, ifelse(m$outcome == "draw", m$draw_p, m$away_p))
  fp <- ifelse(m$outcome == "home", m$fair_home, ifelse(m$outcome == "draw", m$fair_draw, m$fair_away))
  ll_model <- -mean(log(pmin(pmax(mp, eps), 1))); ll_market <- -mean(log(pmin(pmax(fp, eps), 1)))
  cat(sprintf("  3-way log-loss (lower is better): Elo model %.4f · de-vigged closer %.4f → %s by %.4f\n",
              ll_model, ll_market, if (ll_market < ll_model) "the closer is sharper" else "the model is sharper",
              abs(ll_model - ll_market)))
  rows[[length(rows) + 1]] <- data.frame(outcome = "logloss", bin = "model", n = nrow(m), pred = ll_model,
                                         obs = ll_market, obs_lo = NA, obs_hi = NA, stringsAsFactors = FALSE)
  bind_rows(rows)
}

# ---------------------------------------------------------------- D
replay <- function(res, hfa) {
  r <- new.env(hash = TRUE)
  n <- nrow(res); e_out <- numeric(n)
  k_col <- ifelse(grepl("friendly", res$league), ELO_K_FRIENDLY, ELO_K)
  h <- res$home_id; a <- res$away_id; hs <- res$home_score; as_ <- res$away_score; neu <- res$neutral
  for (i in seq_len(n)) {
    rh <- if (is.null(r[[h[i]]])) ELO_START else r[[h[i]]]
    ra <- if (is.null(r[[a[i]]])) ELO_START else r[[a[i]]]
    e <- 1 / (1 + 10 ^ (-(rh - ra + (if (neu[i]) 0 else hfa)) / 400))
    e_out[i] <- e
    s <- if (hs[i] > as_[i]) 1 else if (hs[i] < as_[i]) 0 else 0.5
    gd <- abs(hs[i] - as_[i])
    mult <- if (gd <= 1) 1 else if (gd == 2) 1.5 else (11 + gd) / 8
    d <- k_col[i] * mult * (s - e)
    r[[h[i]]] <- rh + d; r[[a[i]]] <- ra - d
  }
  e_out
}

three_way <- function(e, draw_base) {
  pd <- draw_base * 4 * e * (1 - e)
  ph <- pmax(0, e - pd / 2); pa <- pmax(0, 1 - e - pd / 2)
  s <- ph + pd + pa
  list(ph = ph / s, pd = pd / s, pa = pa / s)
}

section_d <- function(res) {
  n <- nrow(res); start <- as.integer(n * HOLDOUT_FRAC)
  idx <- (start + 1):n
  out <- ifelse(res$home_score > res$away_score, "home", ifelse(res$home_score < res$away_score, "away", "draw"))[idx]
  cat(sprintf("\nD. Refit ELO_HFA x DRAW_BASE on the results table — %s matches, scoring the last %s (3-way log-likelihood per match, higher is better)\n",
              format(n, big.mark = ","), format(n - start, big.mark = ",")))
  rows <- list()
  for (hfa in HFA_GRID) {
    e <- replay(res, hfa)[idx]
    for (db in DRAW_GRID) {
      tw <- three_way(e, db)
      p <- ifelse(out == "home", tw$ph, ifelse(out == "draw", tw$pd, tw$pa))
      rows[[length(rows) + 1]] <- data.frame(hfa = hfa, draw_base = db, loglik = mean(log(pmin(pmax(p, 1e-9), 1))))
    }
  }
  fit <- bind_rows(rows)
  best <- fit[which.max(fit$loglik), ]
  live <- fit[fit$hfa == ELO_HFA_LIVE & fit$draw_base == DRAW_BASE_LIVE, ]
  cat(sprintf("  %5s  %s   (DRAW_BASE)\n", "HFA", paste(sprintf("%8.2f", DRAW_GRID), collapse = "")))
  for (hfa in HFA_GRID) {
    row <- fit[fit$hfa == hfa, ] %>% arrange(draw_base)
    cat(sprintf("  %5.0f  %s\n", hfa, paste(sprintf("%8.4f", row$loglik), collapse = "")))
  }
  cat(sprintf("  best: HFA %.0f, DRAW_BASE %.2f (loglik %.4f); live: HFA %.0f, DRAW_BASE %.2f (loglik %.4f); gap %.4f per match\n",
              best$hfa, best$draw_base, best$loglik, ELO_HFA_LIVE, DRAW_BASE_LIVE, live$loglik, best$loglik - live$loglik))
  fit
}

bets <- load_soccer_bets()
if (nrow(bets) == 0) {
  cat("no settled soccer paper bets yet — run soccer_edge.py --snapshot / --settle (or --backfill)\n")
  quit(status = 0)
}
write.csv(section_a(bets), file.path(OUT_DIR, "soccer_roi.csv"), row.names = FALSE)
write.csv(section_b(bets), file.path(OUT_DIR, "soccer_slices.csv"), row.names = FALSE)
m <- load_soccer_matches()
if (nrow(m) > 0) write.csv(section_c(m), file.path(OUT_DIR, "soccer_calibration.csv"), row.names = FALSE)
res <- load_results()
if (nrow(res) >= 200) write.csv(section_d(res), file.path(OUT_DIR, "soccer_fit.csv"), row.names = FALSE)
cat(sprintf("\nwrote %s\n", file.path(OUT_DIR, "soccer_*.csv")))

File: analysis/05_soccer/soccer_loop.py
"""The soccer analysis loop: does Elo-vs-DraftKings 3-way earn anything, and are the
soccer constants in soccer_edge.py the right priors?

  A. Paper ROI by pick x strength, flat $1, bootstrap 95% CI (twin of 01 for soccer).
  B. Slices a rule can act on: edge band, price band (Wilson CI on hit, flat ROI).
  C. Elo calibration on settled snapshots: binned model P(home) / P(draw) vs observed,
     and log-loss of the model vs the de-vigged closer (is the market sharper?).
  D. Refit on the results table: replay Elo for a grid of ELO_HFA, then score every
     DRAW_BASE by 3-way log-likelihood on the held-out second half. Prints the best
     pair next to the constants the tool is using.

Output: analysis/_out/soccer_roi.csv, soccer_slices.csv, soccer_calibration.csv, soccer_fit.csv.
Mirrors soccer_loop.R — keep in lockstep. Point estimates must match; bootstrap CIs may
differ in the last digit.
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np
import pandas as pd

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from _shared.load_data import wilson
from _shared.load_soccer import load_results, load_soccer_bets, load_soccer_matches

OUT_DIR = Path(__file__).resolve().parents[1] / "_out"
MIN_BETS = 10
MIN_N = 5
BOOT_REPS = 5000
SEED = 20260920

# the constants under test (mirror soccer_edge.py — a change there must be reflected here)
ELO_START, ELO_K, ELO_K_FRIENDLY = 1500.0, 20.0, 10.0
ELO_HFA_LIVE, DRAW_BASE_LIVE = 60.0, 0.26
HFA_GRID = [0.0, 30.0, 60.0, 90.0, 120.0]
DRAW_GRID = [0.20, 0.22, 0.24, 0.26, 0.28, 0.30, 0.32]
HOLDOUT_FRAC = 0.5                    # score the second half of the results table

EDGE_BINS, EDGE_LABELS = [0, 8, 15, 20, 30, 50, 100000], ["0-8", "8-15", "15-20", "20-30", "30-50", "50+"]
PRICE_BINS = [-100000, -200, -110, 99, 150, 200, 250, 100000]
PRICE_LABELS = ["<-200", "-200..-110", "-110..+100", "+100..150", "+150..200", "+200..250", ">+250"]
PROB_BINS, PROB_LABELS = [0, .2, .3, .4, .5, .6, .7, .8, 1.01], ["<.2", ".2-.3", ".3-.4", ".4-.5", ".5-.6", ".6-.7", ".7-.8", ".8+"]


def verdict(lo: float, hi: float) -> str:
    return "PROFITABLE (95% CI > 0)" if lo > 0 else "losing (95% CI < 0)" if hi < 0 else "inconclusive"


def boot_ci(x: np.ndarray, rng: np.random.Generator) -> tuple[float, float]:
    idx = rng.integers(0, len(x), size=(BOOT_REPS, len(x)))
    means = x[idx].mean(axis=1)
    return float(np.percentile(means, 2.5)), float(np.percentile(means, 97.5))


# ---------------------------------------------------------------- A. paper ROI

def section_a(bets: pd.DataFrame, rng) -> pd.DataFrame:
    print(f"A. Paper ROI — {len(bets):,} settled soccer paper bets, backfilled {int(bets['backfill'].sum()):,}, "
          f"flat ROI {100 * bets.pnl_flat.mean():+.1f}%\n")
    groups = [("ALL", "all", bets)]
    groups += [(p, str(s), g) for (p, s), g in bets.groupby(["pick", "strength"])]
    groups += [(p, "any", g) for p, g in bets.groupby("pick")]
    groups += [("ALL", str(s), g) for s, g in bets.groupby("strength")]
    rows = []
    for pick, strength, g in groups:
        if len(g) < MIN_BETS:
            continue
        lo, hi = boot_ci(g.pnl_flat.to_numpy(), rng)
        rows.append({"pick": pick, "strength": strength, "bets": len(g), "wins": int((g.result == "W").sum()),
                     "hit_rate": float((g.result == "W").mean()), "roi_mean": float(g.pnl_flat.mean()),
                     "roi_ci_lo": lo, "roi_ci_hi": hi, "verdict": verdict(lo, hi)})
    out = pd.DataFrame(rows).sort_values("roi_ci_lo", ascending=False)
    print(f"  {'pick':<7}{'str':>4}{'bets':>6}{'wins':>6}{'hit%':>7}{'ROI':>8}{'CI lo':>8}{'CI hi':>8}  verdict")
    for r in out.itertuples():
        print(f"  {r.pick:<7}{r.strength:>4}{r.bets:>6}{r.wins:>6}{100 * r.hit_rate:>6.1f}%{100 * r.roi_mean:>+7.1f}%"
              f"{100 * r.roi_ci_lo:>+7.1f}%{100 * r.roi_ci_hi:>+7.1f}%  {r.verdict}")
    return out


# ---------------------------------------------------------------- B. slices

def slice_table(df: pd.DataFrame, col: str, order: list[str], title: str) -> list[dict]:
    rows = []
    print(f"\n{title}")
    print(f"  {'bucket':<12}{'n':>5}{'W':>5}{'hit%':>8}{'ROI':>8}   95% CI on hit")
    for key in [k for k in order if k in set(df[col])]:
        g = df[df[col] == key]
        if len(g) < MIN_N:
            continue
        w = int((g.result == "W").sum())
        p, lo, hi = wilson(w, len(g))
        rows.append({"dimension": col, "bucket": key, "n": len(g), "wins": w, "hit_rate": p, "hit_lo": lo,
                     "hit_hi": hi, "roi_flat": float(g.pnl_flat.mean())})
        print(f"  {key:<12}{len(g):>5}{w:>5}{100 * p:>7.1f}%{100 * g.pnl_flat.mean():>+7.1f}%   [{100 * lo:.0f}, {100 * hi:.0f}]")
    return rows


def section_b(bets: pd.DataFrame) -> pd.DataFrame:
    b = bets[bets.strength >= 1].copy()
    b["edge_band"] = pd.cut(b.edge, EDGE_BINS, labels=EDGE_LABELS, right=False).astype(str)
    b["price_band"] = pd.cut(b.price, PRICE_BINS, labels=PRICE_LABELS, right=True).astype(str)
    b["strength_s"] = b.strength.astype(str)
    rows = slice_table(b, "edge_band", EDGE_LABELS, "B1. by edge band (does a bigger Elo-vs-DK gap win more?)")
    rows += slice_table(b, "price_band", PRICE_LABELS, "B2. by price band")
    rows += slice_table(b, "pick", ["home", "draw", "away"], "B3. by pick")
    return pd.DataFrame(rows)


# ---------------------------------------------------------------- C. calibration

def section_c(m: pd.DataFrame) -> pd.DataFrame:
    print(f"\nC. Elo calibration on {len(m):,} settled matches with a closer and a rating on both sides")
    rows = []
    for col, label, hit in (("home_p", "P(home win)", "home"), ("draw_p", "P(draw)", "draw"), ("away_p", "P(away win)", "away")):
        print(f"  {label}: model bin -> observed rate (Wilson 95%)")
        band = pd.cut(m[col], PROB_BINS, labels=PROB_LABELS, right=False).astype(str)
        for key in [k for k in PROB_LABELS if k in set(band)]:
            g = m[band == key]
            if len(g) < MIN_N:
                continue
            k = int((g.outcome == hit).sum())
            p, lo, hi = wilson(k, len(g))
            pred = float(g[col].mean())
            rows.append({"outcome": hit, "bin": key, "n": len(g), "pred": pred, "obs": p, "obs_lo": lo, "obs_hi": hi})
            print(f"    {key:<7} n={len(g):>5}  pred {100 * pred:5.1f}%  obs {100 * p:5.1f}% [{100 * lo:5.1f},{100 * hi:5.1f}]  Δ{100 * (p - pred):+5.1f}pp")
    # log-loss: model vs de-vigged closer
    eps = 1e-9
    mp = np.select([m.outcome == "home", m.outcome == "draw"], [m.home_p, m.draw_p], m.away_p).astype(float)
    fp = np.select([m.outcome == "home", m.outcome == "draw"], [m.fair_home, m.fair_draw], m.fair_away).astype(float)
    ll_model, ll_market = -np.log(np.clip(mp, eps, 1)).mean(), -np.log(np.clip(fp, eps, 1)).mean()
    print(f"  3-way log-loss (lower is better): Elo model {ll_model:.4f} · de-vigged closer {ll_market:.4f} "
          f"→ {'the closer is sharper' if ll_market < ll_model else 'the model is sharper'} by {abs(ll_model - ll_market):.4f}")
    rows.append({"outcome": "logloss", "bin": "model", "n": len(m), "pred": ll_model, "obs": ll_market,
                 "obs_lo": np.nan, "obs_hi": np.nan})
    return pd.DataFrame(rows)


# ---------------------------------------------------------------- D. refit

def replay(results: pd.DataFrame, hfa: float) -> np.ndarray:
    """Chronological Elo replay; returns the pre-match expected home score for every row."""
    r: dict[str, float] = {}
    e_out = np.empty(len(results))
    k_col = np.where(results.league.str.contains("friendly"), ELO_K_FRIENDLY, ELO_K)
    for i, (h, a, hs, as_, neu, k) in enumerate(zip(results.home_id, results.away_id, results.home_score,
                                                     results.away_score, results.neutral, k_col)):
        rh, ra = r.get(h, ELO_START), r.get(a, ELO_START)
        e = 1.0 / (1.0 + 10.0 ** (-(rh - ra + (0.0 if neu else hfa)) / 400.0))
        e_out[i] = e
        s = 1.0 if hs > as_ else 0.0 if hs < as_ else 0.5
        gd = abs(hs - as_)
        mult = 1.0 if gd <= 1 else 1.5 if gd == 2 else (11.0 + gd) / 8.0
        d = k * mult * (s - e)
        r[h], r[a] = rh + d, ra - d
    return e_out


def three_way(e: np.ndarray, draw_base: float) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    pd_ = draw_base * 4.0 * e * (1.0 - e)
    ph = np.maximum(0.0, e - pd_ / 2.0)
    pa = np.maximum(0.0, 1.0 - e - pd_ / 2.0)
    s = ph + pd_ + pa
    return ph / s, pd_ / s, pa / s


def section_d(results: pd.DataFrame) -> pd.DataFrame:
    n = len(results)
    start = int(n * HOLDOUT_FRAC)
    out = np.select([results.home_score > results.away_score, results.home_score < results.away_score],
                    ["home", "away"], "draw")[start:]
    print(f"\nD. Refit ELO_HFA x DRAW_BASE on the results table — {n:,} matches, scoring the last {n - start:,} "
          f"(3-way log-likelihood per match, higher is better)")
    rows = []
    for hfa in HFA_GRID:
        e = replay(results, hfa)[start:]
        for db in DRAW_GRID:
            ph, pdr, pa = three_way(e, db)
            p = np.where(out == "home", ph, np.where(out == "draw", pdr, pa))
            ll = float(np.log(np.clip(p, 1e-9, 1)).mean())
            rows.append({"hfa": hfa, "draw_base": db, "loglik": ll})
    fit = pd.DataFrame(rows)
    best = fit.loc[fit.loglik.idxmax()]
    live = fit[(fit.hfa == ELO_HFA_LIVE) & (fit.draw_base == DRAW_BASE_LIVE)].iloc[0]
    print(f"  {'HFA':>5}  " + "".join(f"{db:>8.2f}" for db in DRAW_GRID) + "   (DRAW_BASE)")
    for hfa in HFA_GRID:
        row = fit[fit.hfa == hfa].sort_values("draw_base")
        print(f"  {hfa:>5.0f}  " + "".join(f"{v:>8.4f}" for v in row.loglik))
    print(f"  best: HFA {best.hfa:.0f}, DRAW_BASE {best.draw_base:.2f} (loglik {best.loglik:.4f}); "
          f"live: HFA {ELO_HFA_LIVE:.0f}, DRAW_BASE {DRAW_BASE_LIVE:.2f} (loglik {live.loglik:.4f}); "
          f"gap {best.loglik - live.loglik:.4f} per match")
    return fit


def main() -> None:
    OUT_DIR.mkdir(exist_ok=True)
    rng = np.random.default_rng(SEED)
    bets = load_soccer_bets()
    if bets.empty:
        print("no settled soccer paper bets yet — run soccer_edge.py --snapshot / --settle (or --backfill)")
        return
    section_a(bets, rng).to_csv(OUT_DIR / "soccer_roi.csv", index=False)
    section_b(bets).to_csv(OUT_DIR / "soccer_slices.csv", index=False)
    m = load_soccer_matches()
    if not m.empty:
        section_c(m).to_csv(OUT_DIR / "soccer_calibration.csv", index=False)
    results = load_results()
    if len(results) >= 200:
        section_d(results).to_csv(OUT_DIR / "soccer_fit.csv", index=False)
    print(f"\nwrote {OUT_DIR / 'soccer_*.csv'}")


if __name__ == "__main__":
    main()
